gpiozero/pins/pigpiod.py

```diff
@@ -68,10 +71,65 @@ class PiGPIOPin(Pin):
 
     .. _pigpio: http://abyz.co.uk/rpi/pigpio/
     """
+    def __init__(
+            self, host=os.getenv('PIGPIO_ADDR', 'localhost'),
+            port=int(os.getenv('PIGPIO_PORT', 8888))):
+        super(PiGPIOFactory, self).__init__()
+        self.pin_class = PiGPIOPin
+        self.spi_hardware_class = PiGPIOHardwareSPI
+        self.spi_software_class = PiGPIOSoftwareSPI
+        self.shared_spi_hardware_class = PiGPIOHardwareSPIShared
+        self.shared_spi_software_class = PiGPIOSoftwareSPIShared
+        self._connection = pigpio.pi(host, port)
+        self._host = host
+        self._port = port
+        self._spis = []
 
-    _CONNECTIONS = {} # maps (host, port) to (connection, pi_info)
-    _PINS = {}
+    def close(self):
+        super(PiGPIOFactory, self).close()
+        # We *have* to keep track of SPI interfaces constructed with pigpio;
+        # if we fail to close them they prevent future interfaces from using
+        # the same pins
+        if self.connection:
+            while self._spis:
+                self._spis[0].close()
```

Comment:
Yeah - it is a bit "magic" having SPI interfaces remove themselves from the owning factory's list; the reason for this is that SPI interfaces are basically devices in their own right (i.e. it's conceivable users may create them themselves or use them in their own SPI device classes).

Particularly in the case of pigpio we *must* keep track of these SPI interfaces because if you fail to close remote SPI interfaces, the pigpio daemon refuses to let you create new ones on the same pin combinations (quite reasonably) unless you restart the daemon, of course. Hence why I figured it was best to have the SPI interface deal with this itself rather than leaving it in the hands of the owning device (also reduces code repetition a little, but the daemon state issue is my real reason here).